controllers/configurationpolicy_controller.go

```diff
@@ -1371,97 +1397,171 @@ func (r *ConfigurationPolicyReconciler) determineDesiredObjects(
 		return nil, &scopedGVR, errEvent, nil
 	}
 
+	// Fetch and filter objects using provided objectSelector
 	desiredObjects := []*unstructured.Unstructured{}
+
+	// The object either has no name specified or it is templated in the object definition
+	if desiredName == "" && objectT.ObjectSelector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(objectSelector)
+		if err != nil {
+			log.Error(err, "Failed to select the resources",
+				"objectSelector", fmt.Sprint(objectT.ObjectSelector.String()))
+
+			msg := fmt.Sprintf("Error parsing provided objectSelector: %v", err)
+
+			// only report this error if there wasn't another yet
+			errEvent := &objectTmplEvalEvent{
+				compliant: false,
+				reason:    "objectSelector error",
+				message:   msg,
+			}
+
+			return nil, &scopedGVR, errEvent, err
+		}
+
+		filteredObjects, err := r.DynamicWatcher.List(plc.ObjectIdentifier(), objGVK, "", selector)
+		if err != nil {
+			log.Error(err, "Failed to fetch the resources",
+				"objectSelector", fmt.Sprint(objectT.ObjectSelector.String()))
+
+			msg := fmt.Sprintf("Error listing resources with provided objectSelector: %v", err)
+
+			errEvent := &objectTmplEvalEvent{
+				compliant: false,
+				reason:    "objectSelector error",
+				message:   msg,
+			}
+
+			return nil, &scopedGVR, errEvent, err
+		}
+
+		// Filter objectSelector results using namespaceSelector
+		for _, res := range filteredObjects {
+			resourceNs := res.GetNamespace()
+			// Add the object if the object is in a namespace returned by the namespaceSelector
+			if _, ok := relevantNsNames[resourceNs]; ok {
+				relevantNsNames[resourceNs] = append(relevantNsNames[resourceNs], res.GetName())
+			}
+		}
+	}
+
 	hasTemplate := templates.HasTemplate(objectT.ObjectDefinition.Raw, "", true)
 	needsPerNamespaceTemplating := templateHasObjectNamespaceRegex.Match(objectT.ObjectDefinition.Raw)
 
 	var desiredObj *unstructured.Unstructured
 
-	for _, ns := range relevantNamespaces {
-		var rawDesiredObject []byte
+	for ns, names := range relevantNsNames {
+		for _, name := range names {
```

Comment:
An optimization can be to detect if `ObjectName` is available. If it's not, the template resolution can be skipped after the first name since the template result can't change.